service/portability-service-lambda/portability_service.py

```diff
@@ -0,0 +1,330 @@
+        KeyConditionExpression=Key("EnvironmentId").eq(environment_id)
+    )
+    base_url = ddb_response["Items"][0]["Url"]
+    status_url = os.path.join(base_url, "workflows", workflow_id, "status")
+    headers = ddb_response["Items"][0]["Headers"]
+
+    wes_response = requests.get(status_url, headers=headers)
+    wes_response.raise_for_status()
+
+    return json.loads(wes_response.text)["state"]
+
+
+def environments_post(event, context):
+    """Handle POST to /environments by registering a new external execution environment.
+
+    event is a dict that should have keys
+        - name
+            Name of the environment to use when reporting status, etc.
+        - url
+            Base URL for submitting workflows to the environment
+        - schema
+            Workflow submission schema. Ignored for now, but should be WES once that's
+            finalized.
+        - headers
+            dict of headers to include when making requests to the url
+    """
+
+    environment_id = str(uuid.uuid4())
+
+    environments_table().put_item(
+        Item={
+            "EnvironmentId": environment_id,
+            "Url": event["url"],
+            "Name": event["name"],
+            "Schema": event["schema"],
+            "Headers": event.get("headers", "none")
+        }
+    )
+
+    return {"environment_id": environment_id}
+
+
+def environments_get(event, context):
+    """Handle GET to /environments by returning a list of registered environments."""
+
+    response = environments_table().scan()
+
+    output = []
+    for item in response["Items"]:
+        output.append({"environment_id": item["EnvironmentId"],
+                       "name": item["Name"],
+                       "url": item["Url"],
+                       "schema": item["Schema"]})
+
+    return {"environments": output}
+
+
+def portability_tests_post(event, context):
+    """Handle POST to /portability_tests by submitting workflow requests to registered external
+    environments.
+
+    event is a dict that should have keys
+        - workflow_descriptor (string)
+            The main WDL file to be executed
+        - workflow_params (string)
+            JSON string of inputs to the main WDL
+        - workflow_dependencies (list of strings)
+            List of WDL strings for dependency WDLs
+    """
+
+    # Read all the registered environments so we can iterate over them
+    all_environments = environments_table().scan()["Items"]
+
+    wdl = event["workflow_descriptor"]
+    params = event["workflow_params"]
+
+    test_id = str(uuid.uuid4())
+
+    for environment in all_environments:
+        env_id = environment["EnvironmentId"]
+
+        try:
+            workflow_id = wes_submit(
+                env_id, wdl, params, event.get("workflow_dependencies"))
+            record_test_event(test_id, env_id, "SubmissionSucceeded", workflow_id)
+        except Exception as e:
+            record_test_event(test_id, env_id, "SubmissionFailed", str(e))
```

Comment:
It looks like even if a workflow submission to one (or all) of the registered environments fail, the endpoint will return a 201 created status. Is that the intended behavior here?

Reply:
It is the intended behavior, though I'm open to other suggestions. The idea is that even if all the environments reject your workflow at submission, the portability service still received your workflow and tried to run it. So, you created a test with a status of "Failed" overall and "SubmissionFailed" for some or all environments.

Reply:
Thanks for the clarification here. I think that makes sense!